spyder_kernels/comms/frontendcomm.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+In addition to the remote_call mechanism implemented in CommBase:
+ - Implements _wait_reply, so blocking calls can be made.
+"""
+import os
+import time
+import threading
+
+from spyder_kernels.comms.commbase import CommBase, CommError
+from spyder_kernels.py3compat import TimeoutError, PY2
+
+
+class FrontendComm(CommBase):
+    """Mixin to implement the spyder_shell_api"""
+
+    def __init__(self, kernel):
+        super(FrontendComm, self).__init__()
+
+        # Comms
+        self._pid = os.getpid()
+        self.kernel = kernel
+        self.kernel.comm_manager.register_target(
+            self._comm_name, self._comm_open)
+        self._wait_list = {}
+
+        if not PY2:
+            self._main_thread_id = threading.get_ident()
+
+    def remote_call(self, comm_id=None, blocking=False):
+        """Get a handler for remote calls."""
+        return super(FrontendComm, self).remote_call(
+                blocking=blocking, comm_id=comm_id)
+
+    # --- Private --------
+    def _wait_reply(self, call_id, call_name, timeout):
+        """Wait until the frontend replies to a request."""
+        if call_id in self._call_reply_dict:
+            return
+
+        # There is no get_ident in Py2
+        if not PY2 and self._main_thread_id != threading.get_ident():
+            # We can't call kernel.do_one_iteration from this thread.
+            # And we have no reason to think the main thread is not busy.
+            raise CommError(
+                "Can't make blocking calls from non main threads.")
+        t_start = time.time()
+        self._wait_list[call_id] = call_name
+        while call_id not in self._call_reply_dict:
+            if time.time() > t_start + timeout:
+                raise TimeoutError(
```

Comment:
Could we log this instead of raising an error?

Reply:
That is the same problem than in the other side: What is the return value of a blocking call if the request times out? I think this should be handled by the calling code.

Reply:
I disagree. I think we should simplify things and make this return None.

What could the calling code do with a TimeoutError?

Reply:
Well it could decide to do something else instead. The problem is that this is an interface to call any function. If I call a function and expect that the function might throw an error, I need to add a `try` block, so I can handle the failure gracefully. This is the abstract interface used by all functions. For example, if I call `mydict[key]`, I want to get an exception if there is an error, so I can handle it. I wouldn't want to get a None back (because None might be the return value of a sucessful call.)

Reply:
> What could the calling code do with a TimeoutError?
I mean if we don't want the users to report exceptions on github, why have a windows that ask them to do that? Are the errors happening between spyder-kernel and spyder less important than other errors?

Reply:
The errors should be handled by the calling code as the calling code knows more about the situation than this code. The only things we knpw at this level are: "something went wrong somewhere". The normal way to handle that is to raise an error so the calling code can make a decision about it, not returning a None. Returning a none means that the function doesn't return anything, not that it failed.

Reply:
> If I call a function and expect that the function might throw an error

Exactly, that's *if* you expect an error. If you don't, then the function should handle a None response (which is what I'm suggesting and what we already have in place).

> > What could the calling code do with a TimeoutError?

> I mean if we don't want the users to report exceptions on github, why have a windows that ask them to do that?

That's not the point. The point is that this kind of errors would heavily depend on the current user namespace, which would make them almost unreproducible for us (e.g. what if they start to appear after a week long session in which the user loaded a lot of data and did a lot of things with them?).

In that case, I don't want people to report errors that we won't be able to reproduce nor to fix. What's the point on allowing that?

> The only things we knpw at this level are: "something went wrong somewhere"

Nop, what we know at this point (and what you're reporting with your raise) is "the operation timed out". That's it.

And my proposal is "if things time out, then we should do nothing about it".

Reply:
>Exactly, that's if you expect an error. If you don't, then the function should handle a None response (which is what I'm suggesting and what we already have in place).

Well If the function should handle a `None` response then surely if expects something to not work? I know that this is how things work now but the code was specific to each function. Here as the code is general, and a function might want to know if an error happens, I need to raise an error. For example, if I want to have the other side do something, wait for that to be done, and then proceed, I can call a blocking function. If the blocking function doesn't return anything, I will get a None. Now I have a problem. Did the other side actually finished executing, or was there a timeout? I have no way to know.

> That's not the point. The point is that this kind of errors would heavily depend on the current user namespace, which would make them almost unreproducible for us (e.g. what if they start to appear after a week long session in which the user loaded a lot of data and did a lot of things with them?).

> In that case, I don't want people to report errors that we won't be able to reproduce nor to fix. What's the point on allowing that?

That is specific to the namespace related functions, and should be handled there. This code could be doing anything.

> And my proposal is "if things time out, then we should do nothing about it".

Maybe there is something better to do than nothing, but this function doesn't know that, so it should ask the calling function to figure something out. Maybe the best thing to do is return None? Maybe it is doing something else? The calling function knows about that, but this function doesn't, so the decision should be taken by the calling function.

This is a general interface and could be used by plugins, so there is no way of knowing what this interface will be used for. Assuming that returning None is always the best thing to do is a bit of a stretch. On the other hand, knowing that when you call a blocking function, it might timeout, and plan something seems more reasonable. Again, I am not sugesting that the user should see these errors, I just want the functions that call this interface to handle a timeout.

Reply:
> Well If the function should handle a None response then surely if expects something to not work?

Yes, that's the idea. But your comment made me realize we're proposing two different solutions to the same problem: what to do if a blocking call fails? I propose returning None and you returning a TimeoutError.

That's sort of similar, but then this other comment

> This is a general interface and could be used by plugins, so there is no way of knowing what this interface will be used for.

made me realize that my proposed solution is inferior to yours because in the first case third-party developers would need to know that the response on failure is None, whereas in the second one they would naturally get an error.

So, I defer to your point here and I agree we must return an error and let callers handle it as they consider appropriate.